api/envoy/config/bootstrap/v3/bootstrap.proto

```diff
@@ -260,8 +260,24 @@ message Bootstrap {
   // This may be overridden on a per-cluster basis in cds_config, when
   // :ref:`dns_resolution_config <envoy_v3_api_field_config.cluster.v3.Cluster.dns_resolution_config>`
   // is specified.
+  // *dns_resolution_config* will be deprecated once
+  // :ref:'typed_dns_resolver_config <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.typed_dns_resolver_config>'
+  // is fully supported.
   core.v3.DnsResolutionConfig dns_resolution_config = 30;
 
+  // DNS resolver type configuration extension. This extension can be used to configure c-ares, apple,
+  // or any other DNS resolver types and the related parameters. When c-ares DNS is configured,
+  // :ref:'dns_resolution_config <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.dns_resolution_config>'
+  // will be enclosed in this *typed_dns_resolver_config*.
+  // This configuration is to replace the *dns_resolution_config* configuration eventually.
+  // During the transition period when both *dns_resolution_config* and *typed_dns_resolver_config* exists,
+  // this configuration is optional.
+  // When *typed_dns_resolver_config* is in place, Envoy will use it and ignore *dns_resolution_config*.
+  // When *typed_dns_resolver_config* is missing, the default behavior is in place. Envoy will either use
+  // c-ares DNS library or apple DNS library based on the compiling flag.
+  // [#not-implemented-hide:]
```

Comment:
can we also track the addition of a null DNS configuration mode, where no DNS library is compiled in, and DNS resolutions all fail gracefully?

Reply:
Yes, we can support the NULL DNS configuration mode if the new typed_dns_resolver_config configuration is in place, and check it and have DNS resolutions fail gracefully. 

However, in the case if this configuration is missing, IIUC there is no information for us to identify NULL DNS configuration mode.  Thus, the existing logic (#ifdef) will be executed.

BTW, to make this change backward compatible, this new typed configuration is an optional field for now.